### DISCUSSION TITLE
Loss=nan when training/finetuning conformer model

### DISCUSSION THREAD

**Discussion by whrichd** [NVIDIA-NeMo/NeMo#3006] [Q&A] [answered]
Hi, I've tried to train a conformer model from scratch & finetuning using the pretrained models from NGC. And I seem to be running into `loss=nan` far too often. I don't know if it's a bug or something about my config, or something about conformer in general, so I'm posting this here first, asking for help.

Especially for finetuning, the loss suddenly becomes `nan` after 2-20 iterations with the medium conformer (stt_en_conformer_ctc_medium). The large conformer seems to be stable for longer but I didn't test how long. 

Using the same data and training a medium conformer has worked for me, but not on the first try. I initially encountered `loss=nan` after 7 epochs, and then changed the specaug options, enabled mixed precision training and it went away for no reason and I finished the training. But for finetuning it always fails within the first few iterations. I've tried various learning rates (Noam Annealing, so I've tried 0.5, 1, 2, 5), batch sizes, mixed precision options, specaug (time width from 100 to 0.05), no avail, tried the pretrained citrinet, it works fine.

I tried looking at the intermediate outputs, it seems like the input embeddings all look normal, but at some point on one GPU the encoder output starts to become `nan` for the whole batch which then gets propagated. After a while everything is `nan`. after the encoder for all GPUs.

So from my understanding, it doesn't look like faulty input because in the case of training from scratch it happens after a few epochs. It doesn't look like exploding gradient, I don't see the loss gradually diverging, it happens suddenly. I wonder if it has to do with learning rate or learning rate warm up because if it fails at the very beginning the lr is really really small.

BTW I'm training on 4 A100s, with a private data set.

Has anyone encountered this before? Anyone can help?

**Comment by VahidooX** [comment, marked as answer]
Is it Conformer-CTC or Conformer-Transducer?
We have not tested Conformer fully with mixed precision training. Sometimes loss explodes with mixed precision especially with low weight decay. Does it also happen with fp32?
What is the weight decay and warmup you use? 
What is your lr when that happens?
Have you tried gradient clipping?
By "2-20 iterations" do you mean epochs?

**Reply by whrichd** [reply]
CTC, sorry. I see, I could also try fp32. We have some long data (40-60s) and I was hoping to be able to include those, so use mixed precision to increase batch size. I can try segmenting those audio files.

I’ve tried weight decay of 0.001 and 0, warm up 10000 steps to 20000. 

When it happens for the finetuning, the lr is on the magnitude of about 1*10e-7. When I say 2-20 iterations I mean steps / num of batches, so it’s really at the very beginning of the training…how can I enable gradient clipping?

**Reply by whrichd** [reply]
I tired gradient clipping, didn’t help, but now I’m training with fp32, it looks like it has gone past the point of instability that I saw before…I’ll update if it continues training without issue.

**Reply by VahidooX** [reply]
That would be weird if it happens with lr=1e-7 as actually not much updates are applied to the weight at the early steps. Can you check your manifest file to make sure you do not have audios with size 0 or empty text transcript?
Anyway 40s to 50s is very long to train an ASR model even if mixed precision works. I would suggest trying segmentation or ignore them if they small portion of your dataset.

**Reply by whrichd** [reply]
With mixed precision training turned off it works just fine now. But I see that the improper length distribution having a impact on the training time if not WER. I’ll definitely try to segment them. Thanks!

**Reply by mehadi92** [reply]
@whrichd Does turning off mixed precision fully solve your issue? CC: @VahidooX 

I'm facing the same issue [#6809](https://github.com/NVIDIA/NeMo/discussions/6809)

**Reply by whrichd** [reply]
Hi yeah it fixed the issue. But I think they've since added more support for it, I've been training with BF16 recently.

**Reply by VahidooX** [reply]
If you GPUs support bf16, you may use bf16 and double your batch size per GPU easily as It is safe. 
With fp16, it should be safe in most cases but there can be some exceptions.

**Reply by mehadi92** [reply]
@VahidooX `bf16` fix the NAN loss issue
